pkg/exposer/csi_snapshot.go

```diff
@@ -0,0 +1,386 @@
+				Name:            vs.Name,
+				Namespace:       vs.Namespace,
+				UID:             vs.UID,
+				ResourceVersion: vs.ResourceVersion,
+			},
+			Source: snapshotv1api.VolumeSnapshotContentSource{
+				SnapshotHandle: snapshotVSC.Status.SnapshotHandle,
+			},
+			DeletionPolicy:          snapshotVSC.Spec.DeletionPolicy,
+			Driver:                  snapshotVSC.Spec.Driver,
+			VolumeSnapshotClassName: snapshotVSC.Spec.VolumeSnapshotClassName,
+		},
+	}
+
+	return e.csiSnapshotClient.VolumeSnapshotContents().Create(ctx, vsc, metav1.CreateOptions{})
+}
+
+func (e *csiSnapshotExposer) createBackupPVC(ctx context.Context, ownerObject *unstructured.Unstructured, backupVS string, storageClass string, accessMode string, resource resource.Quantity) (*corev1.PersistentVolumeClaim, error) {
+	backupVCName := ownerObject.GetName()
+
+	volumeMode, err := getVolumeModeByAccessMode(accessMode)
+	if err != nil {
+		return nil, err
+	}
+
+	dataSource := &corev1.TypedLocalObjectReference{
+		APIGroup: &snapshotv1api.SchemeGroupVersion.Group,
+		Kind:     "VolumeSnapshot",
+		Name:     backupVS,
+	}
+
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ownerObject.GetNamespace(),
+			Name:      backupVCName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: ownerObject.GetAPIVersion(),
+					Kind:       ownerObject.GetKind(),
+					Name:       ownerObject.GetName(),
+					UID:        ownerObject.GetUID(),
+					Controller: boolptr.True(),
+				},
+			},
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				corev1.ReadWriteOnce,
+			},
+			StorageClassName: &storageClass,
+			VolumeMode:       &volumeMode,
+			DataSource:       dataSource,
+			DataSourceRef:    dataSource,
+
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource,
+				},
+			},
+		},
+	}
+
+	created, err := e.kubeClient.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(ctx, pvc, metav1.CreateOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "error to create pvc")
+	}
+
+	return created, err
+}
+
+func (e *csiSnapshotExposer) createBackupPod(ctx context.Context, ownerObject *unstructured.Unstructured, backupPVC *corev1.PersistentVolumeClaim, label map[string]string) (*corev1.Pod, error) {
+	podName := ownerObject.GetName()
+
+	var gracePeriod int64 = 0
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      podName,
+			Namespace: ownerObject.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: ownerObject.GetAPIVersion(),
+					Kind:       ownerObject.GetKind(),
+					Name:       ownerObject.GetName(),
+					UID:        ownerObject.GetUID(),
+					Controller: boolptr.True(),
+				},
+			},
+			Labels: label,
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:    podName,
+					Image:   "alpine:latest",
```

Comment:
it's better to specify the ImagePullPolicy: corev1.PullIfNotPresent, or the default image pull policy would be Always

Reply:
Done